doc/developer/design/20231103_privatelink_status_table.md

```diff
@@ -0,0 +1,150 @@
+| `status`           | `text`                     | The status: one of `pending-service-discovery`, `creating-endpoint`, `recreating-endpoint`, `updating-endpoint`, `available`, `deleted`, `deleting`, `expired`, `failed`, `pending`, `pending-acceptance`, `rejected`, `unknown`                        |
+| `occured_at`      | `timestamp with time zone` | The timestamp at which the state change occured.       |
+
+The events in this table will be persisted via storage-managed collections,
+rather than in system tables, so they won't be refreshed and cleared on
+startup. The table columns are modeled after `mz_source_status_history`.
+
+The table will be truncated to only keep a small number of status history
+events per `connection_id` to avoid the table growing forever without bound.
+The truncation will happen on Storage Controller 'start' by leveraging the
+`partially_truncate_status_history` method currently used for truncating
+the source/sink status history tables.
+
+The `CloudResourceController` will expose a `watch_vpc_endpoints` method
+that will establish a Kubernetes `watch` on all `VpcEndpoint`s in the
+namespace and translate them into `VpcEndpointEvent`s (modeled after
+the `watch_services` method on the `NamespacedKubernetesOrchestrator`)
+
+- where `VpcEndpointEvent` is defined as follows:
+   ``` rust
+   struct VpcEndpointEvent {
+      connection_id: GlobalId,
+      status: VpcEndpointState,
+      time: DateTime<Utc>,
+   }
+   ```
+- The `time` field will be determined by inspecting the `Available`
+"condition" on the `VpcEndpointStatus` which contains a `last_transition_time`
+field populated by the VpcEndpoint Controller in the cloud repository.
+- The `status` field will be populated using the `VpcEndpointStatus.state`
+field.
+
+
+The Adapter `Coordinator` (which has a handle to `cloud_resource_controller`)
+will spawn a task on `serve` (similar to where it calls
+`spawn_statement_logging_task`) that calls `watch_vpc_endpoints` to
+receive a stream of `VpcEndpointEvent`s.
+- This task will maintain an in-memory map of the last known state value for
+each connection, compare that to any received `VpcEndpointEvent` event,
+and filter out redundant events.
+- The in-memory map will be initialized based on the last state written to the
+table for each connection. These rows are already read from the table on
+startup in the Storage Controller `partially_truncate_status_history` call,
+which will be refactored to store the `last_n_entries_per_id` it constructs as
+a field on the Storage Controller state, to be consumed by the this task.
+- The task will rate-limit received events using the
+[governor](https://docs.rs/governor/latest/governor/index.html) crate with some
+burst capacity to avoid overloading the coordinator if any endpoint gets stuck
+in a hot fail loop.
+- For each rate-limited batch of events the task will emit a Coordinator
+message `Message::VpcEndpointEvents(BTreeMap<GlobalId, VpcEndpointEvent>)`.
```

Comment:
this all seems good to me! couple of things to be careful with during impl:
- the task can't allow you return an error, it must be permanently running
- it will need some kind of shared access to the catalog so it can begin and start collecting for connections as they are created and dropped
- it will probably be nice to use https://docs.rs/tokio-stream/latest/tokio_stream/struct.StreamMap.html in the impl of this task, message me if you want to discuss in detail!
- i don't think we need to bring the `govenor` crate, we can just have a min interval that we will emit new events to be written to the coordinator, and if multiple events come for the same connection in that interval, we will just choose the last one. i think it would be hard to safely wrap a kube watch with govenor, without possibly causing weird behavior like network timeouts

it would also be nice if this doc clarified _why_ we need to initialize this task with the last event (i assume its entirely to avoid re-emitting the last event?). I assume this will just have the coordinator reach into the storage controller it holds (I think the order of bootstrapping makes this not too bad)

Reply:
for my last question i see now its a resolved open question!

Reply:
> it will need some kind of shared access to the catalog so it can begin and start collecting for connections as they are created and dropped

Nope, it shouldn't! Should be possible to configure the Kubernetes watch to listen for all `VPCEndpoint` events in the namespace, _including_ those that are create or deletion events. Here's the analogous code in the `KubernetesOrchestrator`:

https://github.com/MaterializeInc/materialize/blob/a47c0a852d8c851c54ef4786dd630bc26b796d7e/src/orchestrator-kubernetes/src/lib.rs#L1476-L1488

> it will probably be nice to use https://docs.rs/tokio-stream/latest/tokio_stream/struct.StreamMap.html in the impl of this task, message me if you want to discuss in detail!

(There should only be the _one_ watch, so you shouldn't need to be managing multiple streams.)



> * i don't think we need to bring the `govenor` crate, we can just have a min interval that we will emit new events to be written to the coordinator, and if multiple events come for the same connection in that interval, we will just choose the last one. i think it would be hard to safely wrap a kube watch with govenor, without possibly causing weird behavior like network timeouts

Discussed here: https://github.com/MaterializeInc/materialize/pull/22936#discussion_r1382288410

The reason I proposed using something like `governor` is because I want us to build the muscle of reacting to things in real time. It seems like a shame to build a real time system that has a (say) forced poll interval of 5s. Since the underlying Kubernetes tech here supports proper events, it'd be great to be able to usually bring those events in in real time, but with some protections to prevent overload.

> i think it would be hard to safely wrap a kube watch with govenor, without possibly causing weird behavior like network timeouts

I think it should be fine? Here's the code I had in mind:

```rust
let (coord_tx, coord_rx) = mpsc::channel();
tokio::spawn(async {
    let watch = cloud_resource_controller.watch_vpc_endpoints();
    let rate_limiter = governor::RateLimiter::new(...);
    loop {
        // Wait for an event to become available.
        let event = watch.next().await;
        
        // Wait until we're permitted to tell the coordinator about the event.
        rate_limiter.until_ready().await;
       
        // Drain any additional events from the queue that accumulated while
        // waiting for the rate limiter, deduplicating by VPC endpoint ID.
        let mut events = BTreeMap::new();
        events.insert(event.id, event);
        while let Some(event) = watch.next().now_or_never(); {
            events.insert(event.id, event);
        }

        // Send the event batch to the coordinator.
        coord_tx.send(Message::VpcEndpointEvents(events));
    }
})
```

`StreamExt::next` is cancel safe so no issue there. `watch.next().await` will poll to completion so shouldn't be any issues with network timeouts there. Kubernetes watches are also just long running HTTP requests, so there's no need to poll them on any frequency to keep them alive—and in fact the Kubernetes API server will force terminate them occasionally, to ensure that all clients have good retry behavior.

Reply:
I think @benesch responded to most of the concerns you raised @guswynn, though I'll reword the doc to make it clearer that the kubernetes watch will be on all VPCEndpoints in the namespace, including newly created ones, so I don't think we need to have access to the catalog and don't need to use `StreamMap`.

The `watcher` method in kube-runtime https://docs.rs/kube/latest/kube/runtime/fn.watcher.html also does provide some built-in error handling for us, so retries and connection issues should be handled before we receive events on the stream. 

Re: `governor` - thanks for the code sample @benesch -- I think that approach looks reasonable and if we hit issues with the rate-limiting logic during implementation we can re-evaluate.

Reply:
ah sounds good!